src/py/flwr/cli/log.py

```diff
@@ -0,0 +1,221 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Flower command line interface `log` command."""
+
+import sys
+import time
+from logging import DEBUG, ERROR, INFO
+from pathlib import Path
+from typing import Dict, Optional
+
+import grpc
+import typer
+from typing_extensions import Annotated
+
+from flwr.cli.config_utils import load_and_validate
+from flwr.common.grpc import GRPC_MAX_MESSAGE_LENGTH, create_channel
+from flwr.common.logger import log as logger
+from flwr.proto.exec_pb2 import StreamLogsRequest  # pylint: disable=E0611
+from flwr.proto.exec_pb2_grpc import ExecStub
+
+CONN_REFRESH_PERIOD = 60  # Connection refresh period for log streaming (seconds)
+
+
+def stream_logs(run_id: int, channel: grpc.Channel, duration: int) -> None:
+    """Stream logs from the beginning of a run with connection refresh."""
+    start_time = time.time()
+    stub = ExecStub(channel)
+    req = StreamLogsRequest(run_id=run_id)
+
+    for res in stub.StreamLogs(req):
+        print(res.log_output)
+        if time.time() - start_time > duration:
+            break
+
+
+def print_logs(run_id: int, channel: grpc.Channel, timeout: int) -> None:
+    """Print logs from the beginning of a run."""
+    stub = ExecStub(channel)
+    req = StreamLogsRequest(run_id=run_id)
+
+    while True:
+        try:
+            # Enforce timeout for graceful exit
+            for res in stub.StreamLogs(req, timeout=timeout):
+                print(res.log_output)
+        except grpc.RpcError as e:
+            # pylint: disable=E1101
+            if e.code() == grpc.StatusCode.DEADLINE_EXCEEDED:
+                break
+            if e.code() == grpc.StatusCode.NOT_FOUND:
+                logger(ERROR, "Invalid run_id `%s`, exiting", run_id)
+                break
+            if e.code() == grpc.StatusCode.CANCELLED:
+                break
+
+    channel.close()
+
+
+def log(
+    run_id: Annotated[
+        int,
+        typer.Argument(help="The Flower run ID to query"),
+    ],
+    directory: Annotated[
+        Path,
+        typer.Argument(help="Path of the Flower project to run"),
+    ] = Path("."),
+    federation_name: Annotated[
+        Optional[str],
+        typer.Argument(help="Name of the federation to run the app on"),
+    ] = None,
+    follow: Annotated[
+        bool,
+        typer.Option(
+            "--follow/--no-follow",
```

Comment:
That's a valid point. I initially used the `docker` CLI style, but switched to the `typer` style of `--param/--no-param`. Can see why it's confusing. 

How about this experience:
* `flwr run <APP_DIR> <FEDERATION_NAME>`: Executes run, follows logstream (default)
* `flwr run --silent/-s <APP_DIR> <FEDERATION_NAME>`: Executes run without showing the logstream. 

And when users want to retrieve logs after executing `flwr run ...`:
* `flwr log -- <RUN_ID> <APP_DIR> <FEDERATION_NAME>`: Streams logs (default) if run is still running.
* `flwr log --print/-p -- <RUN_ID> <APP_DIR> <FEDERATION_NAME>`: Print logs.

(Note that `--` is necessary for `flwr log` because of the negative values of `<RUN_ID>`.)

Wdyt @Robert-Steiner (cc @danieljanes @tanertopal)?

Reply:
I like that idea! Maybe `--quite/-q` instead of  `silent`? 
For `log` we can also do it the other way around:
- `flwr log -- <RUN_ID> <APP_DIR> <FEDERATION_NAME>` Print logs.
- `flwr log --follow/-f -- <RUN_ID> <APP_DIR> <FEDERATION_NAME>` Streams logs if run is still running.

is the `--` required if you put the negative value in quotes? e.g. `flwr log "-123" ...`

Reply:
Re: `--quiet/-q`, good idea! It's a consistent pattern with `pip` and `poetry`. I prefer this.

To keep the user experience the same, IMO it's preferable for `flwr log -- ...` to stream the log like in `flwr run <APP_DIR> <FEDERATION_NAME>`. 

IIRC, quotes doesn't work too. Will check.

Reply:
@Robert-Steiner The quotes doesn't work.

Reply:
That is a petty! But thanks for checking it.

Reply:
I will resolve the conversation as the flag design will be discussed elsewhere.